metrics: idiomatic histograms

tbg opened the pull request:
Reworked our usage of histograms. The primary objective was exposing "sane"
histograms to prometheus; prior to this change we exposed ever-changing
streaming quantiles of windowed histograms, which is completely worthless.

To support our in-house time series (which cannot ingest general histograms)
and for "casual" users of histogram information (TxnCoordSender), support for
windowed histograms remains.

In effect, our Histogram type now wraps two internal histograms:
- a "real" histogram which keeps totals and is scraped by prometheus
- a windowed histogram (in turn composed of several subhistograms through which
  it rotates) which is scraped by time series.

This should work reasonably well with the caveat that we cannot sum our
internal time series histograms and expect mathematically meaningful results,
which is avoided for the most part as histogram source aggregation typically
grabs the maximum value observed, which _is_ valid. The notable exception are
rollups, which are not implemented yet but will end up averaging out the same
quantile at different times, giving meaningless results. That latter
restriction seems unimportant in practice given that roll-ups also downsample
dramatically, so the additional loss of fidelity is likely not a concern as
long as we document it appropriately.

Existing tooling was simplified to keep less "useless" histograms. For example,
we were keeping multiple durations of sliding windows for each time series,
but this wasn't used. Now we record only one at a reasonably high resolution
to time series (and prometheus doesn't use the windowed histograms at all).

Similar treatment is possible for "rates" which is another unidiomatic
construct introduced by me back in the day and should be completely removable
(in a future PR).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9810)

<!-- Reviewable:end -->

tamird commented:
:dog:, @mrtracy should look at this. also new stuff doesn't have unit tests?

---

Reviewed 9 of 14 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[kv/txn_coord_sender_test.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVqMsaZtAeVzJSmCI_:-KTVqMsaZtAeVzJSmCIa:b-piatob) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/kv/txn_coord_sender_test.go#L1410)):*
> ```Go
> 	// Metrics lose fidelity, so we can't compare incr directly.
> 	if min := hist.Min(); min < incr-10 {
> 		t.Fatalf("min %d < %d", min, incr)
> ```

update this, too. perhaps extract `10`.

---

*[util/metric/metric.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVw6ogRBpOX-leMXjt:-KTVw6ohg2oJaXnforPC:ba1wjn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L36)):*
> ```Go
> const (
> 	// MaxLatency is the maximum value tracked in latency histograms. Higher
> 	// value will be recorded as this value instead.
> ```

s/value/values/

---

*[util/metric/metric.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVw8_neWL_t2dW_il_:-KTVw8_neWL_t2dW_ila:b-gfmiqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L39)):*
> ```Go
> 	MaxLatency = 10 * time.Second
> 
> 	// Data will leave a latency's windowed histogram after approximately this
> ```

s/latency/latencyRotationInterval/ but even better to rephrase to not mention the name at all (as you do for the very next constant)

---

*[util/metric/metric.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVwtQ2zYqxByiWul65:-KTVwtQ2zYqxByiWul66:b-gixzlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L163)):*
> ```Go
> type Histogram struct {
> 	Metadata
> 	mu         syncutil.Mutex
> ```

anonymous struct pattern?

---

*[util/metric/metric.go, line 167 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVwrGR4Qu5F3ZI-NsC:-KTVwrGR4Qu5F3ZI-NsD:b-7k2bg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L167)):*
> ```Go
> 	sliding    *slidingHistogram
> 
> 	// Immutable fields.
> ```

relic

---

*[util/metric/metric.go, line 191 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVx3yLJvzIzih8YdRF:-KTVx3yLJvzIzih8YdRG:b-rkevj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L191)):*
> ```Go
> func NewLatency(metadata Metadata) *Histogram {
> 	return NewHistogram(
> 		Metadata{
> ```

why can't you just pass `Metadata`? add a comment and consider making a copy and removing the field you're avoiding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-:-KTVxN9C6LNOMfSNsz3u:b-qc98an)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[kv/txn_coord_sender_test.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVqMsaZtAeVzJSmCI_:-KTWY-irX7ibfBOl08jU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/kv/txn_coord_sender_test.go#L1410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> update this, too. perhaps extract `10`.
</details>
Done.

---

*[util/metric/metric.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVw6ogRBpOX-leMXjt:-KTWY5bB0ZkIa1--8Rwu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/value/values/
</details>
Done.

---

*[util/metric/metric.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVw8_neWL_t2dW_il_:-KTWYD1lr7FGMk-vYni-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/latency/latencyRotationInterval/ but even better to rephrase to not mention the name at all (as you do for the very next constant)
</details>
Done.

---

*[util/metric/metric.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVwtQ2zYqxByiWul65:-KTWYewxVdJeKbNp2RNg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L163)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> anonymous struct pattern?
</details>
Done.

---

*[util/metric/metric.go, line 167 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVwrGR4Qu5F3ZI-NsC:-KTWYg44Tr35XxDSIBwy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> relic
</details>
Done.

---

*[util/metric/metric.go, line 191 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVx3yLJvzIzih8YdRF:-KTWZ4KdRqS5I_tqhaU1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/893975c810aaca7d717b40b280c914bb15c71986/util/metric/metric.go#L191)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why can't you just pass `Metadata`? add a comment and consider making a copy and removing the field you're avoiding.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[util/metric/metric.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTWZPJX2IwprTOCXCz0:-KTWZPJX2IwprTOCXCz1:b-mel4ee) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d6a712feacab55c8efca938e0ff9bfcdd29473c/util/metric/metric.go#L39)):*
> a windowed histograms

---

*[util/metric/metric.go, line 176 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTWZckiq_sSbVeZF5It:-KTWZckiq_sSbVeZF5Iu:brm1ya5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d6a712feacab55c8efca938e0ff9bfcdd29473c/util/metric/metric.go#L176)):*
> ```Go
> // precision.
> func NewHistogram(metadata Metadata, duration time.Duration, maxVal int64, sigFigs int) *Histogram {
> 	dHist := newSlidingHistogram(metadata, duration, maxVal, sigFigs)
> ```

is it expected that the metadata is both in this thing and in the Histogram?

---

*[util/metric/sliding_histogram.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTVeMs00DOk7A3bBbMq-r1-21:-KTWZnaSqd8xHtO20wMX:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/247b9aaf7d8778702096b2434c8d65481d50df21/util/metric/sliding_histogram.go#L21)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> whoops, meant to move this file.
</details>
Ack.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-:-KTWZtcOo0iZuXnxE3XT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
metrics: idiomatic histograms

Reworked our usage of histograms. The primary objective was exposing "sane"
histograms to prometheus; prior to this change we exposed ever-changing
streaming quantiles of windowed histograms, which is completely worthless.

To support our in-house time series (which cannot ingest general histograms)
and for "casual" users of histogram information (TxnCoordSender), support for
windowed histograms remains.

In effect, our Histogram type now wraps two internal histograms:

- a "real" histogram which keeps totals and is scraped by prometheus
- a windowed histogram (in turn composed of several subhistograms through which
  it rotates) which is scraped by time series.

This should work reasonably well with the caveat that we cannot sum our
internal time series histograms and expect mathematically meaningful results,
which is avoided for the most part as histogram source aggregation typically
grabs the maximum value observed, which *is* valid. The notable exception are
rollups, which are not implemented yet but will end up averaging out the same
quantile at different times, giving meaningless results. That latter
restriction seems unimportant in practice given that roll-ups also downsample
dramatically, so the additional loss of fidelity is likely not a concern as
long as we document it appropriately.

Existing tooling was simplified to keep less "useless" histograms. For example,
we were keeping multiple durations of sliding windows for each time series,
but this wasn't used. Now we record only one at a reasonably high resolution
to time series (and prometheus doesn't use the windowed histograms at all).

Similar treatment is possible for "rates" which is another unidiomatic
construct introduced by me back in the day and should be completely removable
(in a future PR).

tschottdorf commented:
Review status: 12 of 14 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/metric/metric.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTWZPJX2IwprTOCXCz0:-KTXE1KgBB0511G8ZhkR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d6a712feacab55c8efca938e0ff9bfcdd29473c/util/metric/metric.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > a windowed histograms
</details>
Done.

---

*[util/metric/metric.go, line 176 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810#-KTWZckiq_sSbVeZF5It:-KTW_D2UUo-xauop43MZ:btms36s) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d6a712feacab55c8efca938e0ff9bfcdd29473c/util/metric/metric.go#L176)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it expected that the metadata is both in this thing and in the Histogram?
</details>
Good catch, wasn't actually used any more. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I discussed this change at a high level with @mrtracy earlier today, so I'm going to go ahead with the merge and deploy it.

tamird commented:
Reviewed 1 of 6 files at r2, 2 of 2 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9810)*
<!-- Sent from Reviewable.io -->
